cmd/telemeter-server/main.go

```diff
@@ -275,212 +275,218 @@ func (o *Options) Run() error {
+			_ = s.Shutdown(context.TODO())
+			internalListener.Close()
+		})
+	}
+	{
+		external := http.NewServeMux()
 
-		key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
-		if err != nil {
-			return fmt.Errorf("key generation failed: %v", err)
-		}
+		// v1 routes
+		{
+			var (
+				publicKey  crypto.PublicKey
+				privateKey crypto.PrivateKey
+			)
 
-		privateKey, publicKey = key, key.Public()
-	}
+			if len(o.SharedKey) > 0 {
+				data, err := ioutil.ReadFile(o.SharedKey)
+				if err != nil {
+					return fmt.Errorf("unable to read --shared-key: %v", err)
+				}
 
-	// Configure the whitelist.
-	if len(o.WhitelistFile) > 0 {
-		data, err := ioutil.ReadFile(o.WhitelistFile)
-		if err != nil {
-			return fmt.Errorf("unable to read --whitelist-file: %v", err)
-		}
-		o.Whitelist = append(o.Whitelist, strings.Split(string(data), "\n")...)
-	}
-	for i := 0; i < len(o.Whitelist); {
-		s := strings.TrimSpace(o.Whitelist[i])
-		if len(s) == 0 {
-			o.Whitelist = append(o.Whitelist[:i], o.Whitelist[i+1:]...)
-			continue
-		}
-		o.Whitelist[i] = s
-		i++
-	}
-	whitelister, err := metricfamily.NewWhitelist(o.Whitelist)
-	if err != nil {
-		return err
-	}
+				key, err := loadPrivateKey(data)
+				if err != nil {
+					return err
+				}
 
-	issuer := "telemeter.selfsigned"
-	audience := "telemeter-client"
+				switch t := key.(type) {
+				case *ecdsa.PrivateKey:
+					privateKey, publicKey = t, t.Public()
+				case *rsa.PrivateKey:
+					privateKey, publicKey = t, t.Public()
+				default:
+					return fmt.Errorf("unknown key type in --shared-key")
+				}
+			} else {
+				level.Warn(o.Logger).Log("msg", "Using a generated shared-key")
 
-	jwtAuthorizer := jwt.NewClientAuthorizer(
-		issuer,
-		[]crypto.PublicKey{publicKey},
-		jwt.NewValidator(o.Logger, []string{audience}),
-	)
-	signer := jwt.NewSigner(issuer, privateKey)
+				key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+				if err != nil {
+					return fmt.Errorf("key generation failed: %v", err)
+				}
 
-	external := http.NewServeMux()
-	internal := http.NewServeMux()
+				privateKey, publicKey = key, key.Public()
+			}
 
-	// configure the authenticator and incoming data validator
-	var clusterAuth authorize.ClusterAuthorizer = authorize.ClusterAuthorizerFunc(stub.Authorize)
-	if authorizeURL != nil {
-		clusterAuth = tollbooth.NewAuthorizer(o.Logger, &authorizeClient, authorizeURL)
-	}
+			// Configure the whitelist.
+			if len(o.WhitelistFile) > 0 {
+				data, err := ioutil.ReadFile(o.WhitelistFile)
+				if err != nil {
+					return fmt.Errorf("unable to read --whitelist-file: %v", err)
+				}
+				o.Whitelist = append(o.Whitelist, strings.Split(string(data), "\n")...)
+			}
+			for i := 0; i < len(o.Whitelist); {
+				s := strings.TrimSpace(o.Whitelist[i])
+				if len(s) == 0 {
+					o.Whitelist = append(o.Whitelist[:i], o.Whitelist[i+1:]...)
+					continue
+				}
+				o.Whitelist[i] = s
+				i++
+			}
+			whitelister, err := metricfamily.NewWhitelist(o.Whitelist)
```

Comment:
TODO we need to eventually port this functionality into the remote write api as well

Reply:
We should be able to make this a middleware just like everything else, right?

Reply:
yes exactly